xarray/namedarray/core.py

```diff
@@ -0,0 +1,434 @@
+from __future__ import annotations
+
+import copy
+import math
+import typing
+from collections.abc import Hashable, Iterable, Mapping
+
+import numpy as np
+
+# TODO: get rid of this after migrating this class to array API
+from xarray.core import dtypes
+from xarray.core.indexing import ExplicitlyIndexed
+from xarray.core.utils import Default, _default
+from xarray.namedarray.utils import (
+    T_DuckArray,
+    is_duck_array,
+    is_duck_dask_array,
+    to_0d_object_array,
+)
+
+if typing.TYPE_CHECKING:
+    T_NamedArray = typing.TypeVar("T_NamedArray", bound="NamedArray")
+    DimsInput = typing.Union[str, Iterable[Hashable]]
+    Dims = tuple[Hashable, ...]
+
+
+# TODO: Add tests!
+def as_compatible_data(
+    data: T_DuckArray | np.typing.ArrayLike, fastpath: bool = False
+) -> T_DuckArray:
+    if fastpath and getattr(data, "ndim", 0) > 0:
+        # can't use fastpath (yet) for scalars
+        return typing.cast(T_DuckArray, data)
+
+    # TODO : check scalar
+    if is_duck_array(data):
+        return data
+    if isinstance(data, NamedArray):
+        raise ValueError
+    if isinstance(data, np.ma.MaskedArray):
+        raise ValueError
+    if isinstance(data, ExplicitlyIndexed):
+        # TODO: better that is_duck_array(ExplicitlyIndexed) -> True
+        return typing.cast(T_DuckArray, data)
+
+    # if not isinstance(data, np.ndarray) and (
+    #     hasattr(data, "__array_function__") or hasattr(data, "__array_namespace__")
+    # ):
+    #     return typing.cast(T_DuckArray, data)
+    if isinstance(data, tuple):
+        data = to_0d_object_array(data)
+
+    # validate whether the data is valid data types.
+    return typing.cast(T_DuckArray, np.asarray(data))
+
+
+class NamedArray:
+    __slots__ = ("_dims", "_data", "_attrs")
+
+    def __init__(
+        self,
+        dims: DimsInput,
+        data: T_DuckArray | np.typing.ArrayLike,
+        attrs: dict | None = None,
+    ):
+        self._data: T_DuckArray = as_compatible_data(data)
```

Comment:
No fastpath in as_compatible_data?

Could be a slow down here in Variable. Because we run both `variable.as_compatible_data` and `namedarray.as_compatible_data`

Reply:
good catch! should be fixed now.